### ISSUE TITLE
Include container/conda info in .command.sh

### ISSUE THREAD

**Issue by alanhoyle** [nextflow-io/nextflow#1763]
## New feature

When Nextflow runs in a container or conda environment, include a comment or notation in the .command.sh to reveal the run context.
## Usage scenario 

We primarily run our nextflow workflows using singularity containers.  In the work directories, the `.command.sh` does not contain any reference to the container context and the only way to extract it is to know how the original thing was actually run is to search the `.command.run` for the contents of the `nxf_launch()` function.  

``` bash
$ cat .command.sh
#!/bin/bash -ue
command/to_run
$ grep -A2 'nxf_launch()' .command.run
nxf_launch() {
    set +u; env - PATH="$PATH" SINGULARITYENV_TMP="$TMP" SINGULARITYENV_TMPDIR="$TMPDIR" SINGULARITYENV_NXF_DEBUG=${NXF_DEBUG:=0} singularity exec -B sing_options /path/to/singularity.img /bin/bash -c "cd $PWD; eval $(nxf_container_env); /bin/bash /path/to/NXF_WORK/aa/xxxxxx/.command.run nxf_trace"
}
```

## Suggest implementation 

In the .command.sh add a comment with the container/run context like the one below:  

``` bash
#!/bin/bash -ue
# singularity exec -B sing_options /path/to/singularity.img /bin/bash 
command/to_run
```

A potential problem with this would be that nextflow have to detect the language the process is running using heuristics on the `#!` line and be aware of the "single line comment" notation for the language and use that to inject the comment into the script.  For example, while many of these scripting languages use `#`, if someone used inline groovy in the `"""..."""` section, a `//` should be injected instead.  

```
#!/usr/bin/env groovy
// 
``` 

Alternatively, the run context doc injection could be made to only occur if there wasn't a `#!` line in the `"""..."""` section, since bash would invariably be the thing run in that case.

**Comment by pditommaso** [comment]
I see your point but as mentioned this would corrupt non-Bash scripts unless some extra logic is added to detect the scripting lang and comment syntax, which I think is overkill and would make the task execution mechanism less general.

**Comment by alanhoyle** [comment]
According to [Scripts a la carte documentation](https://www.nextflow.io/docs/latest/process.html#scripts-a-la-carte), all non-bash scripts within the nextflow context would all begin with the `#!`, right?  

I think a good compromise implementation would be to include the container or conda info if the script does not begin with the [shebang.](https://en.wikipedia.org/wiki/Shebang_(Unix)) After all, you're already detecting the presence of that and suppressing the `#!/bin/bash -ue` if it's there.

**Comment by pditommaso** [comment]
Yeah, but this is a Linux (Unix?) convention  https://en.wikipedia.org/wiki/Shebang_(Unix)

You can have any scripting language in the first and only the first line starts with a `#!` prefix. Any following line the `#` is interpreted by the target interpreted throwing an error for languages not using `#` as a comment.

**Comment by alanhoyle** [comment]
It's a unix convention.  I haven't run Nextflow on non-unixes, but it works on Linux and macOS.  

It appears that nextflow is already detecting whether or not there's a shebang in the """...""" section, and if it's lacking it injects `#!/bin/bash -ue`at the top of the .command.sh that gets built for the process.  Am I wrong?  that's the behavior I see and what's documented as linked above.  Is is possible to make Nextflow use a different default interpreter?  I thought we had to have bash running for Nextflow to work.  

My proposal is:  If Nextflow is injecting a shebang line, we know that we're running the bash interpreter in a unix context.  When it injects a shebang, we can safely inject a comment below the shebang to include the run context, be it singularity, docker, conda, or nothing and make it easier to test things when a process fails.  If we are NOT injecting a shebang line, we should not do this as we might not know the proper single-line comment format for the interpreted language.

**Comment by pditommaso** [comment]
Nextflow process can use any scripting interpreter specifying it via Shebang.  `#!/bin/bash -ue` is just the default. 

If you were using `#!/bin/python` the `#` would break the script.

**Comment by alanhoyle** [comment]
@pditommaso:  How about this:  for any .command.sh scripts that Nextflow generates where it inserts the default shebang line (which we know is /bin/bash), if the process is running in a container or a conda, nextflow adds a line that includes the container or conda context.  If the default shebang is NOT used, it is not inserted.    

For example, imagine my workflow has this process like this that used the default interpreter:

```
process in_bash {
    script:
    """
    echo hello world from bash
    """
}

process in_r {
    script:
    """
    #! /bin/env Rscript
    cat('hello world from R')
    """
}


```

If I'm running `-with-singularity /path/to/ubuntu.simg`, the `.command.sh` for `in_bash` would be something like:

```
#!/bin/bash -ue
# singularity exec /path/to/ubuntu.simg bash 

echo hello world from bash
```

For process `in_r`, since nextflow knows it isn't inserting the default shebang, it would be:

```
#! /bin/env Rscript
cat('hello world from R')
```
